### PR TITLE
perf: use pinned host copies in Python hot paths

### PR DESCRIPTION
**Pull request by JacoCheung**
## Summary
- Use pinned host buffers when Python hot paths materialize CUDA tensors on CPU.
- Use pinned host buffers when batch shuffling builds CUDA partition tensors from CPU KK results.
- Reuse CPU workload metadata to avoid an extra CUDA scalar read when computing padded batch sizes.

## Tests
- `python -m py_compile examples/commons/distributed/batch_shuffler.py corelib/dynamicemb/dynamicemb/batched_dynamicemb_function.py`
- `pytest -q examples/tests/commons/test_batch_shuffler_factory.py`
- `torchrun --standalone --nproc_per_node=1 -m pytest -q examples/tests/commons/test_distributed.py -k "batch_allgather or all2all_vs_allgather"`
- CUDA smoke test for the pinned-copy helpers and batch-shuffler partition helper.

Note: DynamicEmb CUDA extension tests require an extension built from the same checkout. The available installed extension used an older `segmented_unique_cuda(keys, table_ids, ...)` signature, so that local failure was not treated as a signal for this Python-only change.

### CODE REVIEW

**Comment by JacoCheung** [comment]
:x: **Pipeline #51298930** -- failed

| Job | Status | Log |
|-----|--------|-----|
| pre_check | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277807) |
| train_build | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277808) |
| inference_build | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277809) |
| tritonserver_build | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277810) |
| build_whl | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277811) |
| dynamicemb_test_fwd_bwd_8gpus | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277812) |
| dynamicemb_test_load_dump_8gpus | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277813) |
| unit_test_1gpu_a100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277814) |
| unit_test_1gpu_h100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277815) |
| unit_test_4gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277816) |
| unit_test_tp_4gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277817) |
| L20_unit_test_1gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277818) |
| inference_unit_test_1gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277819) |
| inference_test_1gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/318277820) |

**Result: 0/14 jobs passed**

[View full pipeline](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/51298930)

File: corelib/dynamicemb/dynamicemb/batched_dynamicemb_function.py [hunk 1]
```diff
@@ -45,6 +45,32 @@
 )
 
 
+def _copy_cuda_tensor_to_pinned_cpu(tensor: torch.Tensor) -> torch.Tensor:
+    if not tensor.is_cuda:
+        return tensor.detach().cpu()
+    cpu = torch.empty(
```
Comment (shijieliu) [inline corelib/dynamicemb/dynamicemb/batched_dynamicemb_function.py]: will this allocation block stream?

Comment (JacoCheung) [reply corelib/dynamicemb/dynamicemb/batched_dynamicemb_function.py]: I think this depends on the PyTorch pinned allocator.
By analogy with ordinary CUDA tensor allocations, the caching allocator takes a buffer from a memory pool, and pinned memory should also have its own memory pool.

**Comment by JacoCheung** [comment]
Cherry-picked into junzhang/benchmark_cont as part of the benchmark workflow branch; closing this standalone PR.
